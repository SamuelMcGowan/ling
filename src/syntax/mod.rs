pub mod lexer;
pub mod parser;
pub mod token;

pub mod source;

use crate::chunk::Chunk;

#[derive(Debug)]
pub(crate) struct ParseContext<'a> {
    pub(crate) source: &'a str,
    pub(crate) chunk: &'a mut Chunk,
}

impl<'a> ParseContext<'a> {
    pub fn new(source: &'a str, chunk: &'a mut Chunk) -> Self {
        Self { source, chunk }
    }
}
