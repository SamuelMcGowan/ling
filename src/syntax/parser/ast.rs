use macro_rules_attribute::{derive, derive_alias};
use ustr::Ustr;

use crate::constants::ConstIdx;

derive_alias! {
    #[derive(Node!)] = #[derive(
        Debug,
        Clone,
        serde::Serialize,
        serde::Deserialize,
    )];
}

#[derive(Node!)]
pub(crate) struct Module {
    pub items: Vec<Item>,
}

#[derive(Node!)]
pub(crate) enum Item {
    Func(Func),
}

#[derive(Node!)]
pub(crate) struct Func {
    pub ident: Ident,
    pub params: Vec<(Ident, Ty)>,
    pub ret_ty: Ty,
    pub body: Expr,
}

#[derive(Node!)]
pub(crate) enum Ty {
    Constructed { ident: Ident, params: Vec<Ty> },
    Unit,
}

#[derive(Node!)]
pub(crate) enum Stmt {
    Expr(Expr),
}

#[derive(Node!)]
pub(crate) enum Expr {
    Const(ConstIdx),
    Unit,
}

#[derive(Node!)]
pub(crate) enum Ident {
    Unresolved(Ustr),
}

impl Ident {
    pub fn unresolved(&self) -> Option<Ustr> {
        match self {
            Self::Unresolved(s) => Some(*s),
        }
    }
}
