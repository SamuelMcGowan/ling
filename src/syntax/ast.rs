use macro_rules_attribute::{derive, derive_alias};
use ustr::Ustr;

use crate::constants::ConstIdx;
use crate::symbol_table::SymbolId;

use super::parser::ParseResult;
use super::source::Span;

derive_alias! {
    #[derive(Node!)] = #[derive(
        Debug,
        Clone,
        serde::Serialize,
    )];
}

pub(crate) struct Spanned<T> {
    pub(crate) inner: T,
    pub(crate) span: Option<Span>,
}

impl<T> Spanned<ParseResult<T>> {
    pub fn transpose(self) -> ParseResult<Spanned<T>> {
        self.inner.map(|inner| Spanned {
            inner,
            span: self.span,
        })
    }
}

#[derive(Node!)]
pub(crate) struct Module {
    pub items: Vec<Item>,
}

#[derive(Node!)]
pub(crate) enum Item {
    Func(Func),
    Dummy,
}

#[derive(Node!)]
pub(crate) struct Func {
    pub ident: Ident,
    pub params: Vec<(Ident, Ty)>,
    pub ret_ty: Ty,
    pub body: Block,
}

#[derive(Node!)]
pub(crate) enum Ty {
    Constructed { ident: Ident, params: Vec<Ty> },
    Unit,
}

#[derive(Node!)]
pub(crate) struct Block {
    pub stmts: Vec<Stmt>,
    pub final_expr: Expr,
}

#[derive(Node!)]
pub(crate) enum Stmt {
    Expr(Expr),

    Loop(Block),
    WhileLoop { cond: Expr, block: Block },

    Declaration { lhs: Ident, rhs: Expr },
    Assignment { lhs: Var, rhs: Expr },

    Dummy,
}

impl Stmt {
    pub fn expect_delim(&self) -> bool {
        match self {
            Self::Expr(expr) => expr.expect_delim(),

            Self::Loop(_) | Self::WhileLoop { .. } => false,

            Self::Declaration { .. } => true,
            Self::Assignment { .. } => true,

            // the recovery function should have recovered past any semicolons
            Self::Dummy => false,
        }
    }
}

#[derive(Node!)]
pub(crate) enum Expr {
    Const(ConstIdx),

    UnaryOp {
        op: UnaryOp,
        expr: Box<Expr>,
    },

    BinOp {
        op: BinOp,
        lhs: Box<Expr>,
        rhs: Box<Expr>,
    },

    Call {
        callee: Box<Expr>,
        args: Vec<Expr>,
    },

    If {
        branches: Vec<IfBranch>,
        else_: Option<Box<Block>>,
    },

    Var(Var),

    Unit,

    Dummy,
}

impl Expr {
    pub fn expect_delim(&self) -> bool {
        !matches!(self, Self::If { .. })
    }
}

#[derive(Node!, Copy)]
pub enum UnaryOp {
    Neg,
    Not,
}

#[derive(Node!, Copy)]
pub enum BinOp {
    LogicalAnd,
    LogicalOr,

    Equal,
    NotEqual,
    Gt,
    Lt,
    GtEqual,
    LtEqual,

    Add,
    Sub,
    Mul,
    Div,
    Mod,

    Pow,

    Call,
    Access,
}

#[derive(Node!)]
pub(crate) struct IfBranch {
    pub(crate) cond: Expr,
    pub(crate) then: Block,
}

#[derive(Node!)]
pub(crate) enum Var {
    Simple(Ident),
    Field { expr: Box<Expr>, field: Ident },
}

#[derive(Node!)]
pub(crate) enum Ident {
    Unresolved(Ustr),
    Resolved(SymbolId),
}

impl Ident {
    pub fn unresolved(&self) -> Option<Ustr> {
        match self {
            Self::Unresolved(s) => Some(*s),
            Self::Resolved(_) => None,
        }
    }
}
