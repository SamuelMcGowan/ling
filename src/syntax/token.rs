use super::span::Span;
use crate::chunk::ConstIdx;

#[derive(Debug, Clone, Copy)]
pub(crate) struct Token {
    pub span: Span,
    pub kind: TokenKind,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub(crate) enum TokenKind {
    Punct(Punct),
    Keyword(Keyword),

    Ident(Ustr),
    Const(ConstIdx),

    Error(Error),
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub(crate) enum Punct {
    LParen,
    RParen,
    LBrace,
    RBrace,
    LBracket,
    RBracket,

    Dot,
    Comma,
    Colon,
    Semicolon,

    Add,
    Sub,
    Mul,
    Div,
    Mod,
    Pow,

    AddEqual,
    SubEqual,
    MulEqual,
    DivEqual,
    ModEqual,
    PowEqual,

    Gt,
    Lt,
    Equal,

    GtEqual,
    LtEqual,
    EqualEqual,

    FatArrow,
    Arrow,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub(crate) enum Keyword {
    Func,
    Return,

    Type,
    Enum,
    Data,
    Typeclass,
    Impl,

    Let,
    Match,

    If,
    Elif,
    Else,

    Loop,
    While,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub(crate) enum Error {
    UnexpectedChar(char),

    InvalidEscape,
    UnterminatedString,

    IntegerOverflow,
    MissingDigits,
}

macro_rules! tkind {
    (punct $punct:ident) => {{
        use crate::syntax::token::*;
        TokenKind::Punct(Punct::$punct)
    }};
    (kwd $kwd:ident) => {{
        use crate::syntax::token::*;
        TokenKind::Keyword(Keyword::$kwd)
    }};

    (ident $ident:expr) => {{
        use ustr::Ustr;
        use crate::syntax::token::*;

        TokenKind::Ident(Ustr::from($ident))
    }};
    (constant $constant:literal) => {{
        use crate::syntax::token::*;
        use crate::chunk::ConstIdx;

        TokenKind::Const(ConstIdx($constant))
    }};

    (error $error:ident $($value:tt)?) => {{
        use crate::syntax::token::*;
        TokenKind::Error(Error::$error $($value)?)
    }}
}

pub(crate) use tkind;
use ustr::Ustr;
