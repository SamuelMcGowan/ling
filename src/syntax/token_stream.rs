use serde::Serialize;

use super::lexer::Lexer;
use super::token::{Bracket, BracketKind, Token, TokenKind};

struct SpareBracket {
    kind: BracketKind,
    token: Token,
}

#[derive(Debug, Clone, Serialize)]
pub(crate) enum TokenTree {
    Token(Token),
    Group {
        bracket_kind: BracketKind,
        tokens: TokenStream,
    },
}

#[derive(Debug, Clone, Serialize)]
pub(crate) struct TokenStream(Vec<TokenTree>);

impl TokenStream {
    pub fn iter(&self) -> impl Iterator<Item = &TokenTree> {
        self.0.iter()
    }
}

pub(crate) fn build_token_stream(lexer: Lexer) -> (TokenStream, Vec<Token>) {
    TokenStreamBuilder::new(lexer).build()
}

struct TokenStreamBuilder<'a> {
    lexer: Lexer<'a>,
    mismatched_brackets: Vec<Token>,
}

impl<'a> TokenStreamBuilder<'a> {
    fn new(lexer: Lexer<'a>) -> Self {
        Self {
            lexer,
            mismatched_brackets: vec![],
        }
    }

    fn build(mut self) -> (TokenStream, Vec<Token>) {
        let mut nodes = vec![];

        while let Some(token) = self.lexer.next() {
            let (node, spare_bracket) = self.parse_node(token);

            if let Some(node) = node {
                nodes.push(node);
            }

            // Nobody claimed this poor bracket :(
            if let Some(spare_bracket) = spare_bracket {
                // we don't emit a node for unmatched closing brackets
                self.mismatched_brackets.push(spare_bracket.token);
            }
        }

        (TokenStream(nodes), self.mismatched_brackets)
    }

    fn parse_node(&mut self, token: Token) -> (Option<TokenTree>, Option<SpareBracket>) {
        match token.kind {
            TokenKind::Bracket(Bracket::Opening(kind)) => {
                // ooh look, a tree
                let (nodes, spare_bracket) = self.parse_tree(token, kind);
                (
                    Some(TokenTree::Group {
                        bracket_kind: kind,
                        tokens: TokenStream(nodes),
                    }),
                    spare_bracket,
                )
            }
            TokenKind::Bracket(Bracket::Closing(kind)) => {
                // an unexpected closing bracket! aargh!
                (None, Some(SpareBracket { kind, token }))
            }
            _ => (Some(TokenTree::Token(token)), None),
        }
    }

    fn parse_tree(
        &mut self,
        token: Token,
        kind: BracketKind,
    ) -> (Vec<TokenTree>, Option<SpareBracket>) {
        let mut nodes = vec![];

        let spare_bracket = loop {
            // get a token
            let Some(inner_token) = self.lexer.next() else {
                // unexpected end of input
                self.mismatched_brackets.push(token);
                break None;
            };

            // check for our closing bracket
            if matches!(inner_token.kind, TokenKind::Bracket(Bracket::Closing(k)) if k == kind) {
                break None;
            }

            // parse a child node
            let (node, spare_bracket) = self.parse_node(inner_token);

            if let Some(node) = node {
                nodes.push(node);
            }

            // check the spare bracket
            if let Some(spare_bracket) = spare_bracket {
                if spare_bracket.kind == kind {
                    // our closing bracket was found!
                    break None;
                } else {
                    // not our bracket
                    // our bracket was unterminated too, so we report that
                    self.mismatched_brackets.push(token);
                    break Some(spare_bracket);
                }
            }
        };

        (nodes, spare_bracket)
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_ron_snapshot;

    use super::*;

    fn get_tree(source: &str) -> (TokenStream, Vec<Token>) {
        let lexer = Lexer::new(source);
        let tt_builder = TokenStreamBuilder::new(lexer);
        tt_builder.build()
    }

    #[test]
    fn simple_tests() {
        assert_ron_snapshot!("token_tree_0", get_tree("( { ) } ( hello )"));
        assert_ron_snapshot!("token_tree_1", get_tree("{{{}}}"));
        assert_ron_snapshot!("token_tree_2", get_tree("((("));
        assert_ron_snapshot!("token_tree_3", get_tree(")))"));
        assert_ron_snapshot!("token_tree_4", get_tree("]{}["));

        assert_ron_snapshot!(
            "token_tree_func",
            get_tree(
                "
            func foo(a: uint, b: uint) -> uint {
                a + b
            }
        "
            )
        );
    }
}
