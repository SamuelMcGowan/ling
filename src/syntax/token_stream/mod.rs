mod builder;

use serde::Serialize;

use super::lexer::Lexer;
use super::token::{BracketKind, Token};

use self::builder::build_token_stream;

#[derive(Debug, Clone, Serialize)]
pub(crate) struct TokenStream(Vec<TokenTree>);

impl TokenStream {
    pub fn from_lexer(lexer: Lexer) -> (Self, Vec<Token>) {
        build_token_stream(lexer)
    }

    pub fn iter(&self) -> impl Iterator<Item = &TokenTree> {
        self.0.iter()
    }
}

#[derive(Debug, Clone, Serialize)]
pub(crate) enum TokenTree {
    Token(Token),
    Group {
        bracket_kind: BracketKind,
        tokens: TokenStream,
    },
}

#[cfg(test)]
mod tests {
    use insta::assert_ron_snapshot;

    use super::*;

    fn get_stream(source: &str) -> (TokenStream, Vec<Token>) {
        TokenStream::from_lexer(Lexer::new(source))
    }

    #[test]
    fn simple_tests() {
        assert_ron_snapshot!("token_tree_0", get_stream("( { ) } ( hello )"));
        assert_ron_snapshot!("token_tree_1", get_stream("{{{}}}"));
        assert_ron_snapshot!("token_tree_2", get_stream("((("));
        assert_ron_snapshot!("token_tree_3", get_stream(")))"));
        assert_ron_snapshot!("token_tree_4", get_stream("]{}["));

        assert_ron_snapshot!(
            "token_tree_func",
            get_stream(
                "
            func foo(a: uint, b: uint) -> uint {
                a + b
            }
        "
            )
        );
    }
}
