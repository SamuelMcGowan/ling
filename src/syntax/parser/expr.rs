use crate::syntax::ast::*;
use crate::syntax::token::{tkind, BracketKind, TokenKind};
use crate::syntax::token_stream::TokenTree;

use super::{ParseError, ParseResult, Parser};

#[repr(u8)]
#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
enum Prec {
    Lowest = 0,

    LogicalOr,
    LogicalAnd,

    Equality,
    Comparison,

    Term,
    Factor,
    Exp,

    Unary,
}

impl BinOp {
    fn prec(&self) -> Prec {
        match self {
            Self::LogicalOr => Prec::LogicalOr,
            Self::LogicalAnd => Prec::LogicalAnd,

            Self::Equal | Self::NotEqual => Prec::Equality,
            Self::Gt | Self::Lt | Self::GtEqual | Self::LtEqual => Prec::Comparison,

            Self::Add | Self::Sub => Prec::Term,
            Self::Mul | Self::Div | Self::Mod => Prec::Factor,
            Self::Pow => Prec::Exp,
        }
    }

    fn r_assoc(&self) -> bool {
        matches!(self, Self::Pow)
    }
}

enum RhsRule {
    BinOp(BinOp),
    Call,
}

impl RhsRule {
    fn should_parse_in_prec(&self, prec: Prec) -> bool {
        match self {
            Self::BinOp(op) => {
                let op_prec = op.prec();
                let r_assoc = op.r_assoc();

                op_prec > prec || r_assoc && op_prec == prec
            }
            Self::Call => true,
        }
    }
}

impl<'a> Parser<'a> {
    pub fn parse_expr(&mut self) -> ParseResult<Expr> {
        self.parse_prec(Prec::Lowest)
    }

    fn parse_prec(&mut self, prec: Prec) -> ParseResult<Expr> {
        let mut expr = self.parse_lhs()?;

        while let Some(tt) = self.tokens.peek() {
            let Some(rhs_rule) = self.get_rhs_rule(tt) else {
                break;
            };

            if !rhs_rule.should_parse_in_prec(prec) {
                break;
            }

            let tt = self.tokens.next().unwrap();

            expr = match rhs_rule {
                RhsRule::BinOp(op) => {
                    let rhs = self.parse_prec(op.prec())?;
                    Expr::BinOp {
                        op,
                        lhs: Box::new(expr),
                        rhs: Box::new(rhs),
                    }
                }

                RhsRule::Call => {
                    let TokenTree::Group { tokens, .. } =  tt else {
                        unreachable!();
                    };

                    let mut parser = self.parser_for_tokens(tokens.into_iter());

                    let args =
                        parser.parse_list("arguments", tkind!(punct Comma), Parser::parse_expr)?;

                    Expr::Call {
                        callee: Box::new(expr),
                        args,
                    }
                }
            };
        }

        Ok(expr)
    }

    fn parse_lhs(&mut self) -> ParseResult<Expr> {
        match self.tokens.next() {
            Some(TokenTree::Token(token)) => match token.kind {
                TokenKind::Const(idx) => Ok(Expr::Const(idx)),
                TokenKind::Ident(ident) => Ok(Expr::Var(Ident::Unresolved(ident))),

                kind if kind == tkind!(punct Sub) => {
                    let expr = self.parse_prec(Prec::Unary)?;
                    Ok(Expr::UnaryOp {
                        op: UnaryOp::Neg,
                        expr: Box::new(expr),
                    })
                }

                kind if kind == tkind!(punct Bang) => {
                    let expr = self.parse_prec(Prec::Unary)?;
                    Ok(Expr::UnaryOp {
                        op: UnaryOp::Not,
                        expr: Box::new(expr),
                    })
                }

                _ => Err(ParseError::unexpected(
                    "an expression",
                    Some(TokenTree::Token(token)),
                )),
            },

            Some(TokenTree::Group {
                bracket_kind: BracketKind::Round,
                tokens,
                ..
            }) => {
                if tokens.is_empty() {
                    Ok(Expr::Unit)
                } else {
                    let mut parser = self.parser_for_tokens(tokens.into_iter());
                    let expr = parser.parse_expr()?;
                    parser.expect_end("parenthesised expression")?;
                    Ok(expr)
                }
            }

            other => Err(ParseError::unexpected("an expression", other)),
        }
    }

    fn get_rhs_rule(&self, tt: &TokenTree) -> Option<RhsRule> {
        Some(match tt {
            TokenTree::Token(t) => {
                let bin_op = match t.kind {
                    t if t == tkind!(kwd LogicalOr) => BinOp::LogicalOr,
                    t if t == tkind!(kwd LogicalAnd) => BinOp::LogicalAnd,

                    t if t == tkind!(punct EqualEqual) => BinOp::Equal,
                    t if t == tkind!(punct BangEqual) => BinOp::NotEqual,

                    t if t == tkind!(punct Gt) => BinOp::Gt,
                    t if t == tkind!(punct Lt) => BinOp::Lt,
                    t if t == tkind!(punct GtEqual) => BinOp::GtEqual,
                    t if t == tkind!(punct LtEqual) => BinOp::LtEqual,

                    t if t == tkind!(punct Add) => BinOp::Add,
                    t if t == tkind!(punct Sub) => BinOp::Sub,

                    t if t == tkind!(punct Mul) => BinOp::Mul,
                    t if t == tkind!(punct Div) => BinOp::Div,
                    t if t == tkind!(punct Mod) => BinOp::Mod,
                    t if t == tkind!(punct Pow) => BinOp::Pow,

                    _ => return None,
                };

                RhsRule::BinOp(bin_op)
            }

            TokenTree::Group {
                bracket_kind: BracketKind::Round,
                ..
            } => RhsRule::Call,

            _ => return None,
        })
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_ron_snapshot;

    use crate::syntax::parser::test_parse;

    #[test]
    fn l_assoc() {
        assert_ron_snapshot!(test_parse("1 + 2 + 3", |p| p.parse_expr()));
    }

    #[test]
    fn r_assoc() {
        assert_ron_snapshot!(test_parse("1 ^ 2 ^ 3", |p| p.parse_expr()));
    }

    #[test]
    fn mul_add() {
        assert_ron_snapshot!(test_parse("1 * 2 + 3", |p| p.parse_expr()));
    }

    #[test]
    fn add_mul() {
        assert_ron_snapshot!(test_parse("1 + 2 * 3", |p| p.parse_expr()));
    }

    #[test]
    fn mul_pow() {
        assert_ron_snapshot!(test_parse("1 * 2 ^ 3", |p| p.parse_expr()));
    }

    #[test]
    fn pow_mul() {
        assert_ron_snapshot!(test_parse("1 ^ 2 * 3", |p| p.parse_expr()));
    }
}
