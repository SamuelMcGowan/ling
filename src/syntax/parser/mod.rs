pub mod expr;
pub mod item;

use super::ast::Spanned;
use super::source::Span;
use super::token::{BracketKind, Token, TokenKind};
use super::token_stream::{TokenIter, TokenTree};

#[derive(Debug, Clone, serde::Serialize)]
pub(crate) enum ParseError {
    Unexpected {
        expected: String,
        found: Option<Span>,
    },
    InvalidAssignmentTarget(Span),
    InvalidImplicitReturn(Span),
}

impl ParseError {
    fn unexpected(expected: impl Into<String>, found: Option<TokenTree>) -> Self {
        Self::Unexpected {
            expected: expected.into(),
            found: found.map(|tt| tt.span()),
        }
    }
}

pub(crate) type ParseResult<T> = Result<T, ParseError>;

pub(crate) struct Parser<'a> {
    errors: &'a mut Vec<ParseError>,
    tokens: TokenIter,
}

// Utilities.
impl<'a> Parser<'a> {
    pub fn new(tokens: TokenIter, errors: &'a mut Vec<ParseError>) -> Self {
        Self { tokens, errors }
    }

    fn parser_for_tokens(&mut self, tokens: TokenIter) -> Parser {
        Parser {
            errors: self.errors,
            tokens,
        }
    }

    fn eat_kind(&mut self, kind: TokenKind) -> bool {
        match self.tokens.peek() {
            Some(TokenTree::Token(token)) if token.kind == kind => {
                self.tokens.next();
                true
            }
            _ => false,
        }
    }

    fn expect_kind(&mut self, kind: TokenKind) -> ParseResult<Token> {
        match self.tokens.next() {
            Some(TokenTree::Token(token)) if token.kind == kind => Ok(token),
            other => Err(ParseError::unexpected(
                format!("token of kind {kind:?}"),
                other,
            )),
        }
    }

    fn expect_group(
        &mut self,
        kind: BracketKind,
        expected: impl Into<String>,
    ) -> ParseResult<TokenIter> {
        match self.tokens.next() {
            Some(TokenTree::Group {
                bracket_kind,
                tokens,
                ..
            }) if bracket_kind == kind => Ok(tokens.into_iter()),
            other => Err(ParseError::unexpected(expected.into(), other)),
        }
    }

    fn expect_end(&mut self, tree_name: &str) -> ParseResult<()> {
        match self.tokens.next() {
            None => Ok(()),
            Some(tree) => Err(ParseError::unexpected(
                format!("end of {tree_name}"),
                Some(tree),
            )),
        }
    }

    fn eat_group(&mut self, kind: BracketKind) -> Option<TokenIter> {
        match self.tokens.peek() {
            Some(TokenTree::Group { bracket_kind, .. }) if *bracket_kind == kind => {
                let Some(TokenTree::Group { tokens, ..}) = self.tokens.next() else {
                    unreachable!();
                };

                Some(tokens.into_iter())
            }
            _ => None,
        }
    }

    fn parse_list<T>(
        &mut self,
        tree_name: &str,
        delim: TokenKind,
        parse: impl Fn(&mut Self) -> ParseResult<T>,
    ) -> ParseResult<Vec<T>> {
        let mut items = vec![];

        while !self.tokens.at_end() {
            let item = parse(self)?;
            items.push(item);

            if !self.eat_kind(delim) {
                if let Some(token) = self.tokens.next() {
                    return Err(ParseError::unexpected(
                        format!("token of kind {delim:?} or end of {tree_name:?}"),
                        Some(token),
                    ));
                }
                break;
            }
        }

        Ok(items)
    }
}

// Error handling/recovery.
impl Parser<'_> {
    fn parse_or_recover<T>(
        &mut self,
        parser: impl Fn(&mut Self) -> ParseResult<T>,
        recover: impl Fn(&mut Self) -> T,
    ) -> T {
        parser(self).unwrap_or_else(|error| {
            self.report(error);
            recover(self)
        })
    }

    fn parse_spanned<T>(&mut self, parser: impl Fn(&mut Self) -> T) -> Spanned<T> {
        let start = self.tokens.peek().map(|tt| tt.span());
        let inner = parser(self);
        let end = self.tokens.prev_span();

        let span = match (start, end) {
            (Some(start), Some(end)) => Some(start.union(end)),
            _ => None,
        };

        Spanned { inner, span }
    }

    fn recover_until(&mut self, kinds: &[TokenKind]) {
        while let Some(tt) = self.tokens.peek() {
            if matches!(tt, TokenTree::Token(token) if kinds.contains(&token.kind)) {
                break;
            }
            self.tokens.next();
        }
    }

    fn recover_past(&mut self, kind: TokenKind) {
        self.tokens
            .by_ref()
            .take_while(|token| !matches!(token, TokenTree::Token(token) if token.kind == kind))
            .for_each(|_| {});
    }

    fn report(&mut self, error: ParseError) {
        self.errors.push(error);
    }
}

#[cfg(test)]
pub(crate) fn test_parse<T>(
    source: &str,
    f: impl Fn(&mut Parser) -> T,
) -> (T, Vec<Token>, Vec<ParseError>) {
    use crate::syntax::lexer::Lexer;
    use crate::syntax::token_stream::TokenStream;

    let lexer = Lexer::new(source);
    let (tokens, mismatched_brackets) = TokenStream::from_lexer(lexer);

    let mut errors = vec![];
    let mut parser = Parser::new(tokens.into_iter(), &mut errors);
    let res = f(&mut parser);

    (res, mismatched_brackets, errors)
}
