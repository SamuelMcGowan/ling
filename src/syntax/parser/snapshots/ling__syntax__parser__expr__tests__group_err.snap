---
source: src/syntax/parser/expr.rs
expression: "test_parse(\"(1 1)\", |p| p.parse_expr())"
---
(Err(Unexpected(
  expected: "end of parenthesised expression",
  found: Some(Token(Token(
    span: Span(
      start: 3,
      end: 4,
    ),
    kind: Const(ConstIdx(1)),
  ))),
)), [], [])
